tensorflow/lite/micro/kernels/if.cc

```diff
@@ -76,24 +121,65 @@ TfLiteStatus Prepare(TfLiteContext* context, TfLiteNode* node) {
       context, num_outputs,
       graph_info.NumSubgraphOutputs(op_data->then_subgraph_index));
 
+  op_data->intermediate_copy_buffer = context->AllocatePersistentBuffer(
+      context, CalculateIntermediateBufferSize(context, node));
+
   return kTfLiteOk;
 }
 
-TfLiteStatus Eval(TfLiteContext* context, TfLiteNode* node) {
+TfLiteStatus CopyOutputBetweenIfAndSubgraph(
+    TfLiteContext* context, TfLiteNode* node, MicroGraph& graph_info,
+    int active_branch_subgraph_index, CopyOutputDirection copy_direction) {
   const OpData* op_data = reinterpret_cast<OpData*>(node->user_data);
+  int8_t* intermediate_copy_buffer =
+      reinterpret_cast<int8_t*>(op_data->intermediate_copy_buffer);
+  for (size_t i = 0;
+       i < graph_info.NumSubgraphOutputs(active_branch_subgraph_index); ++i) {
+    const TfLiteEvalTensor* output =
+        tflite::micro::GetEvalOutput(context, node, i);
 
-  const TfLiteTensor* cond;
-  TF_LITE_ENSURE_OK(context, GetInputSafe(context, node, 0, &cond));
-  bool cond_value = cond->data.b[0];
+    TfLiteEvalTensor* subgraph_output =
+        graph_info.GetSubgraphOutput(active_branch_subgraph_index, i);
 
-  tflite::MicroContext* micro_context = tflite::GetMicroContext(context);
-  MicroGraph& graph_info = micro_context->graph();
+    // These checks must occur in Eval since TfLiteEvalTensors are not available
```

Comment:
While the eval tensors are not available, equivalent TfLiteTensors are available during prepare. I would slightly prefer to keep this logic within prepare, since they are designed to ensure correct model formation.

Another way to put it - the only way this check could catch something not caught during prepare would be if our eval tensors have a different type or shape from the tensors in the model and during prepare, which would indicate a much larger issue with TFLM overall.

Reply:
I also prefer to do this in prepare. Unfortunately, the check here requires getting input/output tensor of other branch subgraphs and I think we don't have a good way to get those. The way to get tensor during Prepare is  GetInputSafe or GetOutput , which requires TfLiteNode as input.  I feel it may be too much to put in new api just for this purpose. Feel free to let me know if there are better ways to get tensor.